api/document/json_importer/importer.py

```diff
@@ -0,0 +1,131 @@
+from collections import defaultdict
+from typing import (Callable, DefaultDict, Dict, List, NamedTuple, Optional,
+                    Type)
+
+from django.db import transaction
+
+from document.models import Annotation, DocNode, ExternalLink, FootnoteCitation
+from document.tree import JSONAwareCursor, JsonDict
+from reqs.models import Policy
+
+
+class PreparedNode(NamedTuple):
+    node: JsonDict
+    content: List[JsonDict]
+    children: List[JsonDict]
+
+    @classmethod
+    def from_json(cls, node):
+        prepped = {}
+        for key in PRIMITIVE_DOC_NODE_FIELDS:
+            if key in node:
+                prepped[key] = node[key]
+        content = node.get('content', [])
+        prepped['text'] = get_content_text(content)
+        return cls(
+            node=prepped,
+            content=content,
+            children=node.get('children', []),
+        )
+
+
+PRIMITIVE_DOC_NODE_FIELDS = [
+    # TODO: Include 'identifier' here?
+    'node_type',
+    'type_emblem',
+    'marker',
+    'title',
+]
+
+Annotator = Callable[
+    [JSONAwareCursor, JsonDict, int],
+    Annotation
+]
+
+annotators: Dict[str, Annotator] = {}
+
+
+def annotator(fn: Annotator):
+    annotators[fn.__name__] = fn
+    return fn
+
+
+@annotator
+def footnote_citation(cursor: JSONAwareCursor, content: JsonDict,
+                      start: int) -> FootnoteCitation:
+    text = content['text']
+    referencing = list(cursor.filter(
+        lambda m: m.node_type == 'footnote'
+        and m.type_emblem == text.strip()
+    ))
+    if not referencing:
+        raise ValueError(f'unable to find footnote for citation {text}')
+    return FootnoteCitation(
+        doc_node=cursor.model, start=start, end=start + len(text),
+        footnote_node=referencing[0].model,
+    )
+
+
+@annotator
+def external_link(cursor: JSONAwareCursor, content: JsonDict,
+                  start: int) -> ExternalLink:
+    return ExternalLink(
+        doc_node=cursor.model, start=start,
+        end=start + len(content['text']), href=content['href']
+    )
+
+
+AnnotationDict = DefaultDict[Type[Annotation], List[Annotation]]
+
+
+def derive_annotations(cursor: JSONAwareCursor) -> AnnotationDict:
+    annotations: AnnotationDict = defaultdict(list)
+    start = 0
+    for content in cursor.json_content:
+        content_type = content['content_type']
+        if content_type != '__text__':
+            if content_type not in annotators:
+                raise ValueError(f"no annotator found for {content_type}")
+            anno = annotators[content_type](cursor, content, start)
+            annotations[anno.__class__].append(anno)
+        start += len(content['text'])
+
+    for child_cursor in cursor.children():
+        for cls, annos in derive_annotations(child_cursor).items():
+            annotations[cls].extend(annos)
+
+    return annotations
+
+
+def get_content_text(content: List[JsonDict]):
+    return ''.join([c['text'] for c in content])
```

Comment:
The linter may complain here: there's no need for the list; `join`'ll work on the generator.